---

**<font color="green">Комментарий ревьюера</font>**

Привет Катерина, меня зовут Толя! Я ревьюер твоего проекта.

Поздравляю с первым выполненным проектом. Ты проделала огромную работу, нашла много интересных инсайтов и сделал несколько контринтуитивных выводов, молодец. :)

По ходу тетрадки ты найдешь комментарии к проекту, они будут в отдельных ячейках Markdown с заголовком «Комментарий ревюера» + номер итерации проверки (v2, v3 ...). Постарайся учесть эти комментарии для дальнейших проектов.

Пожалуйста, не удаляй комментарии.

---



<div class="alert alert-block alert-danger">
    

### Комментарий ревьюера

Такой комментарий нужно исправить обязательно, он критически влияет на удачное выполнение проекта.
</div>
    
---

<div class="alert alert-block alert-warning">
    

### Комментарий ревьюера


Такой комментарий является рекомендацией или советом. Можешь использовать их на своё усмотрение.
</div>

---

<div class="alert alert-block alert-success">
    

### Комментарий ревьюера

Такой комментарий  говорит о том, что было сделано что-то качественное и правильное :)
</div>

---

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
#Для анализа "надежности заемщиков" проводим анализ данных, в котором проверим наличие ошибок, неточных данных, соответствие типов данных друг другу.
# При наличии ошибок - принимаем решение как их устранить- ШАГ 2. При наличии дубликаотов - удалить - ШАГ 3.
# Далее проводи лемматизацию целей кредита, для того, чтобы в дальнейшем посмотреть - влияет ли "цель кредита" на его "возвратность"- раздел "Лемматизация"
# Соберем данные и категоризируем цели кредита в зависимости от назначения. ШАГ "Категоризация"
# Проверим - насколько влияет Семейное положение,Наличие и количество детей и Ежемесячный доход на возврат кредита. 
# Данные действия будем проводить с помощью сводных таблиц метод - pivot.
# Делаем Заключительные Выводы.


import pandas as pd
data = pd.read_csv('/datasets/data.csv')
data.head(30)



children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   
20         1   -1311.604166         36              среднее             1   
21         1    -253.685166         33              среднее             1   
22         1   -1766.644138         24              среднее             1   
23         0    -272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0    -529.191635         28               высшее             0   
28         1    -717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0         женат / замужем                 0      F    сотрудник     0   
1         женат / замужем                 0      F    сотрудник     0   
2         женат / замужем                 0      M    сотрудник     0   
3         женат / замужем                 0      M    сотрудник     0   
4        гражданский брак                 1      F    пенсионер     0   
5        гражданский брак                 1      M    компаньон     0   
6         женат / замужем                 0      F    компаньон     0   
7         женат / замужем                 0      M    сотрудник     0   
8        гражданский брак                 1      F    сотрудник     0   
9         женат / замужем                 0      M    сотрудник     0   
10        женат / замужем                 0      M    компаньон     0   
11        женат / замужем                 0      F    сотрудник     0   
12       гражданский брак                 1      M    пенсионер     0   
13        женат / замужем                 0      F    сотрудник     0   
14       гражданский брак                 1      F    компаньон     1   
15        женат / замужем                 0      F    сотрудник     0   
16        женат / замужем                 0      F    сотрудник     0   
17       гражданский брак                 1      M    сотрудник     0   
18         вдовец / вдова                 2      F    пенсионер     0   
19              в разводе                 3      F    сотрудник     0   
20        женат / замужем                 0      M    сотрудник     0   
21       г

### Вывод

Дата_сет имеет текст с разной величиной букв:образование указано и с большой и с маленькой буквы, а также с использавнием разных регистров. 
Имеются отрицательные значения days_employed и total_income. Возможно - возникли при введении "вручную". 
Имеются пропущенные значения в колонке Трудовой стаж . Возможно сотрудник никогда не работал, но тогда в графе тип занятости тоже было бы пусто: значит пропущено значение при заполнении.


 Шаг 2. Предобработка данных

Обработка пропусков


In [2]:
data.info()  #смотрим данные
data['days_employed'].value_counts() 
print(data['days_employed'].value_counts()) #выясняем количество уникальных значений, чтобы посмотреть тип данных

print(data.isnull().sum()) #считаем кол-во пропусков, чтобы выяснить где есть пустые значения

data['days_employed'] = data['days_employed'].fillna(value='') #заменили Nan на пустое значение для удобства в работе
print(data['days_employed'].head(10)) #проверка

print(data['total_income'].value_counts())
data['total_income'] = data['total_income'].fillna(value='')
print(data['total_income'].head(10)) #проверка

data['children'].value_counts()
print(data['children'].value_counts())

data['dob_years'].value_counts()
print(data['dob_years'].value_counts())

data['family_status_id'].value_counts()
print(data['family_status_id'].value_counts())

data['family_status'].value_counts()
print(data['family_status'].value_counts())

data['education_id'].value_counts()
print(data['education_id'].value_counts())

data['education'].value_counts()
print(data['education'].value_counts())

data['income_type'].value_counts()
print(data['income_type'].value_counts())

data['gender'].value_counts()
print(data['gender'].value_counts())

data['debt'].value_counts()
print(data['debt'].value_counts())

data.loc[data['children'] ==-1, ['children']] = 1
data.loc[data['children'] == 20, ['children']] = 2        
print(data['children'].value_counts())



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
-986.927316     1
-7026.359174    1
-4236.274243    1
-6620.396473    1
-1238.560080    1
               ..
-2849.351119    1
-5619.328204    1
-448.829898     1
-1687.038672    1
-582.538413     1
Name: days_employed, Length: 19351, dtype: int64
children               0
days_employed       2174
dob_years              0
educat

### Вывод

При анализе данных методом info видно несоответствие типов данных в столбцах 'days_employed' и 'total_income'Данные имеют формат object вместо float.   В столбце 'total_income', dtype: object - необходимо поменять тип данных на числовой. Столбец 'days_employed' имеет большое кол-во отрицательных и "длинных значений", причина возникновения- непонятная. Скорее всего - ошибка при внесении данных вручную. Предлагается в дальнейшем не использовать значения данного столбца, а в столбце 'total_income'изменить тип на числовой.
Также столбец 'children'имеет одно отрицательное значение -1 (47 шт - 0,21%)- величина незначительная - можно ей пренебречь, но можно и предположить,что это опечатка и значение=1. Тоже самое относительно количества детей 20 чел.-предполагаем, что верное значение 2 и проводим замену для обоих значений.

---

<div class="alert alert-block alert-warning">
    

### Комментарий ревьюера
Классное решение вывести уникальные значения столбцов.
    
По поводу замены количества детей -- спорное решение, почему -1 == 1, а 20 == 2?

В коде выше часто повторяется одно действие но для разных столбцов:  :
    
`data['family_status_id'].value_counts()
print(data['family_status_id'].value_counts())`
    
Схожие действия очень хорошо оборачивать в функцию.
    

</div>

---

### Замена типа данных

In [3]:

data['total_income'] = pd.to_numeric(data['total_income'], errors='coerce')# меняем тип данных столбца 'total_income' на числовой 

print(data['total_income'])


0        253875.639453
1        112080.014102
2        145885.952297
3        267628.550329
4        158616.077870
             ...      
21520    224791.862382
21521    155999.806512
21522     89672.561153
21523    244093.050500
21524     82047.418899
Name: total_income, Length: 21525, dtype: float64


### Вывод

Теперь с данным столбцом можно проводить математические операции. Выбран данный метод т.к. надо перевести данные из object в integrate.


### Обработка дубликатов

In [4]:
data['education'] = data['education'].str.lower() #приводим символы в строках с типом данных 'object' к нижнему регистру
print(data['education'].value_counts())#проверяем уникальные значения,чтобы удостовериться, что все надписи со строчных букв

data['family_status'] = data['family_status'].str.lower()
print(data['family_status'].value_counts()) #проверяем уникальные значения,чтобы удостовериться, что все надписи со строчных букв

data['income_type'] = data['income_type'].str.lower()
print(data['income_type'].value_counts()) #проверяем уникальные значения,чтобы удостовериться, что все надписи со строчных букв

data['purpose'] = data['purpose'].str.lower()
#print(data['purpose'].value_counts()) #проверяем уникальные значения,чтобы удостовериться, что все надписи со строчных букв

data.duplicated().sum() #смотрим количество дубликатов

data = data.drop_duplicates() #удаляем дубликаты

data.duplicated().sum() #проверяем - удалены ли все дубликаты


среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64
женат / замужем          12380
гражданский брак          4177
не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64
сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64


0

### Вывод

Для поиска дубликатов сначала провели замену всех символов в строках типа 'object', все символы привели к нижнему регистру для того, чтобы они имели одинаковый размер и можно было произвести поиск дубликатов.Далее проверили данное изменение методом value_counts. После этого проверили наличие дубликатов методом df.duplicated().sum()- получилось 71 шт, после чего сделали удаление дубликатов методом drop_duplicates и проверили удаление дубликатов методом df.duplicated().sum(), результат - 0.Можно проводить дальнейшие действия. Дубликаты могли появиться при введении данных, некорректном копировании или слиянии таблиц.

### Лемматизация

In [5]:
from pymystem3 import Mystem
m = Mystem()

text = ' '.join(data.purpose)# т.к.текст находится в колонке purpose используем метод Join

lemmas = m.lemmatize(text)
print(lemmas)

from collections import Counter
print(Counter(lemmas))


['покупка', ' ', 'жилье', ' ', 'приобретение', ' ', 'автомобиль', ' ', 'покупка', ' ', 'жилье', ' ', 'дополнительный', ' ', 'образование', ' ', 'сыграть', ' ', 'свадьба', ' ', 'покупка', ' ', 'жилье', ' ', 'операция', ' ', 'с', ' ', 'жилье', ' ', 'образование', ' ', 'на', ' ', 'проведение', ' ', 'свадьба', ' ', 'покупка', ' ', 'жилье', ' ', 'для', ' ', 'семья', ' ', 'покупка', ' ', 'недвижимость', ' ', 'покупка', ' ', 'коммерческий', ' ', 'недвижимость', ' ', 'сыграть', ' ', 'свадьба', ' ', 'приобретение', ' ', 'автомобиль', ' ', 'покупка', ' ', 'жилой', ' ', 'недвижимость', ' ', 'строительство', ' ', 'собственный', ' ', 'недвижимость', ' ', 'недвижимость', ' ', 'строительство', ' ', 'недвижимость', ' ', 'на', ' ', 'покупка', ' ', 'подержать', ' ', 'автомобиль', ' ', 'на', ' ', 'покупка', ' ', 'свой', ' ', 'автомобиль', ' ', 'недвижимость', ' ', 'приобретение', ' ', 'автомобиль', ' ', 'на', ' ', 'покупка', ' ', 'подержать', ' ', 'автомобиль', ' ', 'сыграть', ' ', 'свадьба', ' ', 'опера

 Вывод

Лемматизация данных целей кредита проведена с помощью библиотеки лемматизации на русском языке pymystem3. Наиболее часто встречающиеся слова,которые можно идентифицировать как "цель кредита как таковая": 1 - приобретение недвижимости - 6351 запросов, покупка автомобиля - 4306 запросов, образование-4013. Меньше всего встречаемость слов - свадьба - 2324, строительство - 1878, ремонт - 607. 

### Категоризация данных

In [6]:
data.loc[:,'lemmas'] = data['purpose'].apply(m.lemmatize)# добавляем лематизированный столбец к столбцу 'purpose'
#print(data.loc[:,'lemmas']) # проверяем                         
str = ' '.join(Mystem().lemmatize(text)) # задаем строку с лемматизированным текстом

def purpose_lemmas(str): # создаем функцию,принимающую на вход "запрос категории цели кредита" и возвращающую значение в лематизированный столбец в "понятном для чтения виде" 
    if str == 'недвижимость':
          return 'приобретение недвижимости'
    if 'недвижимость' in str:
        return 'приобретение недвижимости'
    
    if str == 'жилье':
          return 'приобретение недвижимости'
    if 'жилье' in str:
        return 'приобретение недвижимости'
            
    if str == 'ремонт':
        return 'проведение ремонта'
    if 'ремонт' in str:
        return 'проведение ремонта'
    
    if str == 'строительство':
        return 'строительство недвижимости'
    if 'строительство' in str:
        return 'строительство недвижимости'
    
    if str == 'автомобиль':
        return 'покупка автомобиля'
    if 'автомобиль' in str:
        return 'покупка автомобиля'
     
    if str == 'образование':
         return 'получение образования'
    if 'образование' in str:
        return 'получение образования'
        
    if str == 'свадьба':
        return 'организация свадьбы'
    if 'свадьба' in str:
        return 'организация свадьбы'
    return 0
    
#print(purpose_lemmas('свадьба')) # проверка после написания всей функции    

data.loc[:,'lemmas'] = data.loc[:,'lemmas'].apply(purpose_lemmas)

data.loc[:,'lemmas']


0        приобретение недвижимости
1               покупка автомобиля
2        приобретение недвижимости
3            получение образования
4              организация свадьбы
                   ...            
21520    приобретение недвижимости
21521           покупка автомобиля
21522    приобретение недвижимости
21523           покупка автомобиля
21524           покупка автомобиля
Name: lemmas, Length: 21454, dtype: object

---

<div class="alert alert-block alert-warning">
    

### Комментарий ревьюера


Категоризация верная, подобные функции можно упростить, используя axis=0 или axis=1  в методе apply.
</div>

---

### Вывод

Результат лемматизации показал 6 основные категории "целей кредитования": 1 - приобретение недвижимости - 6351 запросов, покупка автомобиля - 4306 запросов, образование-4013, свадьба - 2324, строительство - 1878, ремонт - 607. 
В соответствии с данными запросами были сформированы категории запросов,согласно наличия "ключевых слов": приобретение недвижимости(сюда отнесены все запросы,имеющие в своем составе запросы "недвижимость и жилье"), покупка автомобиля - все запросы,имеющие в составе слово "автомобиль"), образование, организация свадьбы,строительство и ремонт. Последние 2 категории(строительство и ремонт) не объединялись в одну в связи с тем, что строительство можно применить только к "недвижимости",а ремонт можно провести для "транспортных средств".
Еще одно слово - "операция" - опираясь на семантический анализ - это может быть "операция органов" и "операция с недвижимостью или др.имуществом". Не отнесла ни в одну из категорий, т.к. не могу сформулировать запрос: проведение операции(как написал бы любой потребитель в заявке) - опять таки не дает четкого ответа на вопрос: "Операция с чем?...на чем???" Добавить какие-то доп.слова для поиска - неизвестно какие.



### Шаг 3. Ответьте на вопросы



- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [7]:
#Пишем функцию, возвращающую значение, согласно правилу:
#нет детей,нет невозвратов кредита - "бездетные с хорошей кредитной историей'"
#нет детей,есть невозврат кредита - "бездетные с плохой кредитной историей'"
#1-2 ребенка, не было невозвратов кредита - 'мало детей с хорошей кредитной историей'
#1-2 ребенка, был невозврат кредита - 'мало детей с плохой кредитной историей'
#> 2 детей - многодетные


def vozvrat_debt_children(row):
        
    debt = row['debt']
    children = row['children']

    if debt == 0: #  нет детей и невозвратов
        if children == 0:
            return 'бездетный с хорошей кредитной историей'
        
    if debt == 1: #  нет детей и есть невозврат кредита
        if children == 0:
            return 'бездетный с плохой кредитной историей'    
        

    if debt == 0: #семья, с количеством детей от 1-2 чел
        if 1 >= children <= 2:
            return 'мало детей с хорошей кредитной историей'
        
    if debt == 1:
        if 1 >= children <= 2:
            return 'мало детей с плохой кредитной историей'
        
        
    if debt == 0:
        if 2 > children <= 5:
            return 'многодетные с хорошей кредитной историей'#количество детей от 3-5

    return 'многодетные с плохой кредитной историей'

row_values = [1, 0] #значения наличия "кредитной истории" и количества детей 
row_columns = ['debt', 'children'] #названия столбцов

row = pd.Series(data=row_values, index=row_columns)
#print(vozvrat_debt_children(row)) # проверка работы функции
    
data['debt_children'] = data.apply(vozvrat_debt_children, axis=1)  
dep_debt_children = data['debt_children'].value_counts()

print(dep_debt_children)
print()

print('Анализ зависимости возвратов/невозвратов кредитов в зависимости от наличия детей')

print()
total_witout_child = 13028+1063
print('Общее количество бездетных заемщиков', total_witout_child)
total_with_child = 21525-13028-1063
print('Общее количество заемщиков c детьми',total_with_child)

data_grouped_children = data.groupby('children')[['debt']].agg(['count',"sum", "mean"])# составляем сводную таблицу по "возврату кредита в зависимости отколичества детей"
#data_grouped_children['ratio'] = data_grouped_children["debt"]["sum"]/data_grouped_children["debt"]["count"]

data_grouped_children



бездетный с хорошей кредитной историей     13028
мало детей с хорошей кредитной историей     4410
многодетные с плохой кредитной историей     2508
бездетный с плохой кредитной историей       1063
мало детей с плохой кредитной историей       445
Name: debt_children, dtype: int64

Анализ зависимости возвратов/невозвратов кредитов в зависимости от наличия детей

Общее количество бездетных заемщиков 14091
Общее количество заемщиков c детьми 7434


debt                
          count   sum      mean
children                       
0         14091  1063  0.075438
1          4855   445  0.091658
2          2128   202  0.094925
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000

### Вывод 

Общее количество "заявителей на кредит" -  21520
Целевые группы:
0 - нет детей;
1 - 1 ребенок в семье;
2 - 2 ребенка;
3 - 3 ребенка и т.д.

По результатам таблицы можно сделать вывод, что зависимость есть: соотношение заемщиков с детьми к бездетным - 53%.

Вывод 1: Люди с детьми кредит берут с меньшей интенсивностью почти в 2 раза.

Вывод 2: Бездетные заемщики наиболее "надежные" - количество невозвратных кредитов в данном целевом сегменте всего 7,5%.
Вывод 3: на втором месте по степени надежности - семьи с 3 детьми - количество "невозвратов" - 8,1%.
Вывод 4: Семьи с 1, 2 и 4 детьми - самые "ненадежные" - количество "невозвратов" в данной категории - 9,1%, 9,4%(имеющие 2 детей) и 9,7% (самые "ненадежные", имеющие 4 детей). 
Вывод 5: Многодетные семьи,с 5 детьми кредиты не берут.



---

<div class="alert alert-block alert-success">
    

### Комментарий ревьюера v2

Получилось очень наглядно, молодец. С выводом 5 не совсем согласен. Строго говоря выборка из 9 человек не валидная, но все 9 многодетных заемщиков вернули кредит в срок.
</div>

---

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [8]:
print()

print()
print('Решение через функцию')
print()


def vozvrat_debt_family_st(row):
    ##пишем функцию, которая возвращает значение согласно следующему правилу:
    ###Заявитель с сем. положением "женат / замужем" , не имеющий невозвратов по кредиту - 
    ###статус'женат / замужем с хорошей кредитной историей'
    ###Заявитель с сем. положением "женат / замужем" , имеющий невозвраты по кредиту  - статус 
    ###'женат / замужем с плохой кредитной историей
    ###Заявитель с сем. положением "гражданский брак" , не имеющий невозвратов по кредиту - 
    ###статус 'гражданский брак с хорошей кредитной историей'
    ###Заявитель с сем. положением "гражданский брак" , имеющий невозвраты по кредиту  - статус 
    ###'гражданский брак с плохой кредитной историей'
    
       
     
    debt = row['debt']
    family_status = row['family_status_id']

    if debt == 0: #  женат / замужем нет невозвратов
        if family_status == 0: # женат / замужем
            return 'женат / замужем с хорошей кредитной историей'
        
    if debt == 1: #  есть невозврат кредита
        if family_status == 0: # женат / замужем
            return 'женат / замужем с плохой кредитной историей'    
        

    if debt == 0: #нет невозврата кредита
        if family_status == 1: #гражданский брак
            return 'гражданский брак с хорошей кредитной историей'
        
    if debt == 1: #есть невозврат кредита
        if family_status == 1:
            return 'гражданский брак с плохой кредитной историей'
          
      
    if debt == 0: #нет невозврата кредита
        if family_status == 2: #вдовец / вдова 
            return 'вдовец / вдова с хорошей кредитной историей'
        
    if debt == 1: #есть невозврат кредита
        if family_status == 2:
            return 'вдовец / вдова с плохой кредитной историей'
        
    if debt == 0: #нет невозврата кредита
        if family_status == 3: #в разводе
            return 'в разводе с хорошей кредитной историей'
        
    if debt == 1: #есть невозврат кредита
        if family_status == 3:
            return 'в разводе с плохой кредитной историей'
        
    if debt == 0: #нет невозврата кредита
        if family_status == 4: #не женат / не замужем
            return 'не женат / не замужем с хорошей кредитной историей'
    return 'не женат / не замужем с плохой кредитной историей'    

row_values = [0, 0] #значения наличия "кредитной истории" и "семейного положения"
row_columns = ['debt', 'family_status_id'] #названия столбцов

row = pd.Series(data=row_values, index=row_columns)
#print(vozvrat_debt_family_st(row)) # проверка работы функции
    
data['debt_family_st'] = data.apply(vozvrat_debt_family_st, axis=1)  
dep_debt_family_st = data['debt_family_st'].value_counts()

print(dep_debt_family_st)


# Создаем сводную таблицу через метод groupby                                                  
data_grouped_FS = data.groupby('family_status')[['debt']].agg(['count',"sum", "mean"])
#data_grouped_FS['ratio'] = data_grouped_FS["debt"]["sum"]/data_grouped_FS["debt"]["count"]

data_grouped_FS



Решение через функцию

женат / замужем с хорошей кредитной историей          11408
гражданский брак с хорошей кредитной историей          3763
не женат / не замужем с хорошей кредитной историей     2536
в разводе с хорошей кредитной историей                 1110
женат / замужем с плохой кредитной историей             931
вдовец / вдова с хорошей кредитной историей             896
гражданский брак с плохой кредитной историей            388
не женат / не замужем с плохой кредитной историей       274
в разводе с плохой кредитной историей                    85
вдовец / вдова с плохой кредитной историей               63
Name: debt_family_st, dtype: int64


debt               
                       count  sum      mean
family_status                              
в разводе               1195   85  0.071130
вдовец / вдова           959   63  0.065693
гражданский брак        4151  388  0.093471
женат / замужем        12339  931  0.075452
не женат / не замужем   2810  274  0.097509

### Вывод

   Зависимость между семейным положением и возвратом кредита в срок есть: наиболее надежные заемщики - люди, ставшие ""вдовцом/вдовой", в разводе, семейные люди.
По данным категориям % невозврата 6,5%, 7,1%, 7,5%.
Далее идут: не женат - 9,7%, гражданский брак - 9,3%.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [9]:
#определяем медиану для каждого вида образования, для заполнения пробелов в 'total_income'
mediana_med = data.loc[data['education'] == 'среднее','total_income'].median() #рассчитываем медиану для каждого вида образования, для заполнения пропусков в столбце 'total_income'
mediana_high = data.loc[data['education'] == 'высшее','total_income'].median()

mediana_nf = data.loc[data['education'] == 'неоконченное высшее','total_income'].median()
mediana_beg = data.loc[data['education'] == 'начальное','total_income'].median()

mediana_phd = data.loc[data['education'] == 'ученая степень','total_income'].median()

#print(mediana_med) #проверка медианы
#print(mediana_phd)
#print(data['education'].value_counts()) #проверка

#заполняем пробел в 'total_income' значениями медианы для каждого вида образования

data.loc[data['education']=='среднее','total_income']= data.loc[data['education']=='среднее','total_income'].fillna(mediana_med)
data.loc[data['education']=='высшее','total_income']= data.loc[data['education']=='высшее','total_income'].fillna(mediana_high)

data.loc[data['education']=='неоконченное высшее','total_income']= data.loc[data['education']=='неоконченное высшее','total_income'].fillna(mediana_nf)
data.loc[data['education']=='начальное','total_income']= data.loc[data['education']=='начальное','total_income'].fillna(mediana_beg)

data.loc[data['education']=='ученая степень','total_income']= data.loc[data['education']=='ученая степень','total_income'].fillna(mediana_phd)

#print(data['total_income'].head(30))

mediana_income = data['total_income'].median()
print("Средний уровень дохода по заявителям {:.0f}".format(mediana_income)) #средний доход по колонке 'total_income'

print("Минимальный уровень дохода по заявителям {:.0f}".format(data['total_income'].min())) #находим минимальный доход 

print("Максимальный уровень дохода по заявителям {:.0f}".format(data['total_income'].max())) #находим максимальный доход

print()
print()
print('Зависимость возврата кредита от уровня дохода')
print()


def vozvrat_debt_income(row):
    ##пишем функцию, которая возвращает значение согласно следующему правилу:
    ### при определении категорий опираемся на медиану по уровню дохода, она равна -  141147.
    
    ###Заявитель с очень высоким уровнем дохода > 500000, не имеющий невозвратов по кредиту 
    ###статус'очень высокий уровень дохода с хорошей кредитной историей'
    ###Заявитель с очень высоким уровнем дохода > 500000, имеющий невозвраты по кредиту 
    ###статус'очень высокий уровень дохода с плохой кредитной историей'
    
    ###Заявитель с высоким уровнем дохода > 141147, не имеющий невозвратов по кредиту - 
    ###статус'высокий уровень дохода с хорошей кредитной историей'
    ###Заявитель с высоким уровнем дохода > 141147, имеющий невозвраты по кредиту  - статус
    ###'высокий уровень дохода с плохой кредитной историей
    
    ###Заявитель со средним уровнем дохода  от 93157 до 141147 (нижняя граница - квантиль 0.66 от 141147), не имеющий невозвратов по кредиту - 
    ###статус 'средний уровень дохода с хорошей кредитной историей'
    ###Заявитель со средним уровнем дохода от 93157 до 141147, имеющий невозвраты по кредиту  - статус 
    ###'средний уровень дохода с плохой кредитной историей'
    
    ### Низкий уровень дохода меньше 93157.
       
     
    debt = row['debt']
    total_income = row['total_income']
    
    if debt == 0: #  нет невозвратов
        if total_income > 500000: # уровень очень высокого дохода
            return 'очень высокий уровень дохода с хорошей кредитной историей'
        
    if debt == 1: #  есть невозврат кредита
        if total_income > 500000: # уровень высокого дохода
            return 'очень высокий уровень дохода с плохой кредитной историей'    
        

    if debt == 0: #  нет невозвратов
        if total_income > 141147: # уровень высокого дохода
            return 'высокий уровень дохода с хорошей кредитной историей'
        
    if debt == 1: #  есть невозврат кредита
        if total_income > 141147: # уровень высокого дохода
            return 'высокий уровень дохода с плохой кредитной историей'    
        

    if debt == 0: #  нет невозвратов
        if 93157 < total_income < 141147: # уровень высокого дохода
            return 'средний уровень дохода с хорошей кредитной историей'
        
    if debt == 1: #  есть невозврат кредита
        if 93157 < total_income < 141147: # уровень высокого дохода
            return 'средний уровень дохода с плохой кредитной историей'    
          
      
    if debt == 0: #  нет невозвратов
        if total_income < 93157: # уровень высокого дохода
            return 'низкий уровень дохода с хорошей кредитной историей'
                       
    return 'низкий уровень дохода с плохой кредитной историей'    

row_values = [1, 501000] #значения наличия "кредитной истории" и "уровня доходов"
row_columns = ['debt', 'total_income'] #названия столбцов

row = pd.Series(data=row_values, index=row_columns)
#print(vozvrat_debt_income(row)) # проверка работы функции
    
data['debt_income'] = data.apply(vozvrat_debt_income, axis=1)  
dep_debt_debt_income = data['debt_income'].value_counts()

print(dep_debt_debt_income)

# Пишем функцию для использования в построении сводной таблицы
# Создаем 4 категории дохода:
# 93157 - 'низкий уровень дохода'
# 141147 - 'средний уровень дохода'
# 500 000 и выше - 'очень состоятельный'


def vozvrat_debt_income_1(total_income):
    
    if total_income <=93157: # уровень высокого дохода
        return 'низкий уровень дохода'
     
    if total_income <141147: # уровень высокого дохода
        return 'средний уровень дохода'   
         
    if total_income >=141147: # уровень высокого дохода
        return 'очень высокий уровень' 
       
    if total_income >=500000: # уровень очень высокого дохода
        return 'очень состоятельный' 
                               
    return 'очень состоятельный'
          

print(vozvrat_debt_income_1(94000)) # проверка после написания всей функции    

data.loc[:,'vozvrat_debt_TI'] = data.loc[:,'total_income'].apply(vozvrat_debt_income_1)

data.loc[:,'vozvrat_debt_TI']


data_grouped_TI = data.groupby('vozvrat_debt_TI')[['debt']].agg(['count',"sum", "mean"])
#data_grouped_TI['ratio'] = data_grouped_TI["debt"]["sum"]/data_grouped_TI["debt"]["count"]

data_grouped_TI



Средний уровень дохода по заявителям 141148
Минимальный уровень дохода по заявителям 20667
Максимальный уровень дохода по заявителям 2265604


Зависимость возврата кредита от уровня дохода

высокий уровень дохода с хорошей кредитной историей          9678
средний уровень дохода с хорошей кредитной историей          6437
низкий уровень дохода с хорошей кредитной историей           3390
высокий уровень дохода с плохой кредитной историей            827
средний уровень дохода с плохой кредитной историей            608
низкий уровень дохода с плохой кредитной историей             292
очень высокий уровень дохода с хорошей кредитной историей     208
очень высокий уровень дохода с плохой кредитной историей       14
Name: debt_income, dtype: int64
средний уровень дохода


debt               
                        count  sum      mean
vozvrat_debt_TI                             
низкий уровень дохода    3682  292  0.079305
очень высокий уровень   10727  841  0.078400
средний уровень дохода   7045  608  0.086302

### Вывод

Вывод 1. Наиболее часто кредит берут люди с очень высоким уровнем дохода - выше 500 000. Также они являются наиболее надежными заемщиками - процент возврата по таким кредитным историям - 92,2 %. Всего 7,8 % невозвратов.
Вовод 2. Люди со средним уровнем дохода также надежны - частота возвратов 91,4%
вывод 3: Люди с низким уровнем дохода также возвращают кредит в 92,1 случаях из 100.

Общий вывод для задания - уровень дохода не влияет на частотность возврата кредита, либо разбивка по целевым группам была сделана с не совсем корректными параметрами.


---

<div class="alert alert-block alert-warning">
    

### Комментарий ревьюера v2
    
Вывод `Наиболее часто кредит берут люди с очень высоким уровнем дохода - выше 500 000.` Не совсем верный, так как в функции категоризации `vozvrat_debt_income_1` строка `if total_income >=500000:` ни разу не сработала из-за отсекающего условия `if total_income >=141147` Строго говоря большинство заемщиков с доходом больше 141147

</div>

---

---

<div class="alert alert-block alert-success">
    

### Комментарий ревьюера

Классная идея с заполнением пропусков в зависимости от образования. 
</div>

---

- Как разные цели кредита влияют на его возврат в срок?

In [10]:
data_grouped = data.groupby('lemmas')[['debt']].agg(['count',"sum", "mean"])

data_grouped

debt               
                           count  sum      mean
lemmas                                         
организация свадьбы         2324  186  0.080034
покупка автомобиля          4306  403  0.093590
получение образования       4013  370  0.092200
приобретение недвижимости  10811  782  0.072334

### Вывод

Зависимость между целями кредита и возвратом кредита - хуже всего возращают кредит, который берут под покупку автомобиля 9,4% и получение образования. Далее - "организация свадьбы" - 8% невозвратов и приобретение недвижимости - 7,2%.
Самый "обеспеченный" и возвратный кредит - приобретение недвижимости.

---

<div class="alert alert-block alert-success">
    

### Комментарий ревьюера

Отлично, все предельно ясно и прозрачно. 
    
Столбцы `mean` и `ratio` повторяют друг друга, они несут дополнительный смысл?
</div>

---

### Шаг 4. Общий вывод

Был проанализирован массив данных, содержащий заявки на кредит.
На стадии предобработки данных было выявлено болшое количество пропусков - порядка 10% в уровне дохода, наличие недостоверных данных - отрицательные значения в Продолжительности занятости, "аномальное количество детей" в соответствующей графе(-1 и больше 20).
В результате наличия непригодных для работы данных были сделаны гипотезы по "замене данных с -1 на 1 и 20 на 2 в столбце "Children" Предполагаем, что это опечатки". Пропуски в столбце "Total_income" были заменены на медианной значение, которое было вычислено для соответствующих уровней образования. Данные значения были применены к категориям "Total_income" согласно уровню образования. Также были удалены дубликаты строк.

Рекомендации: Относиться внимательнее к предоставляемым данным - проверять наличие информации во всех столбцах и при наличии пропусков и "недостоверных" данных либо давать рекомендации на что опираться при их заполненнии(целевые группы, статистические данные, средние показатели "по группе" или другому параметру),либо проводить дополнительный сбор информации, чтобы погрешности были меньше.

Были поставлены следующие вопросы:
Есть зависимость между наличием детей и возвратом кредита в срок?
Ответ: 
Бездетные заемщики наиболее "надежные" - количество невозвратных кредитов в данном целевом сегменте всего 8%.

Есть ли зависимость между семейным положением и возвратом кредита всрок?
Ответ: 
Зависимость между семейным положением и возвратом кредита в срок есть: наиболее надежные заемщики - люди, ставшие "вдовцом/вдовой", в разводе, семейные люди.

Вопрос: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
Ответ: Нет, даже очень состоятельные люди "забывают" возвратить кредит в срок. Либо статисткиа по данной категории была собрана слишком давно, когда уровень дохода был низким и были "невозвраты". Либо разбиение на "категории" было некорректным. 

Вопрос:Как разные цели кредита влияют на его возврат в срок?
Ответ: Самый "обеспеченный" и возвратный кредит - приобретение недвижимости.

Рекомендации по выбору "заемщиков" - предпочтение стоит отдавать людям, являющимся "вдовой/вдовцом", в разводе или семейным людям. Также надежные заемщики  - люди, не имеющие детй, либо семьи, имеющие 3 детей.
Зависимости между возвратом кредита и уровнем дохода нет. 


---

<div class="alert alert-block alert-warning">
    

### Комментарий ревьюера

Выводы очень качественные, молодец. Единственно е чего не хватает это твоих рекомендаций относительно скоринговой модели и работе с багами в данных, которые ты предобработала на шаге 2.

</div>

---

---

<div class="alert alert-block alert-success">
    

### Общий комментарий ревьюера v2

Катерина, большое спасибо за проработанные комментарии. Критичных замечаний у меня нет.
    
Что можно улучшить?
    
* По форматированию текста советую посмотреть небольшую статью по `markdown` https://guides.hexlet.io/markdown/, может быть найдешь что то интересное для будущих работ. 
* чтобы инициализировать функции на ходу посмотри небольшую статью про `lambda`  функции https://webdevblog.ru/kak-ispolzovat-v-python-lyambda-funkcii/ так твой код станет еще мощнее.    

Итог ревью -- исследование завершено успешно. Поздравляю со сданным проектом.

Успехов в дальнейшем обучении!

</div>

---

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.